src/main/java/org/skriptlang/skript/bukkit/tags/TagRegistry.java

```diff
@@ -0,0 +1,206 @@
+
+		if (TagModule.PAPER_TAGS_EXIST) {
+			try {
+				List<Tag<Material>> materialTags = new ArrayList<>();
+				for (Field field : MaterialTags.class.getDeclaredFields()) {
+					if (field.canAccess(null))
+						//noinspection unchecked
+						materialTags.add((Tag<Material>) field.get(null));
+				}
+				PaperTagSource<Material> paperMaterialTags = new PaperTagSource<>(materialTags, TagType.BLOCKS, TagType.ITEMS);
+				tagSourceMap.put(TagType.BLOCKS, paperMaterialTags);
+				tagSourceMap.put(TagType.ITEMS, paperMaterialTags);
+
+				List<Tag<EntityType>> entityTags = new ArrayList<>();
+				for (Field field : EntityTags.class.getDeclaredFields()) {
+					if (field.canAccess(null))
+						//noinspection unchecked
+						entityTags.add((Tag<EntityType>) field.get(null));
+				}
+				PaperTagSource<EntityType> paperEntityTags = new PaperTagSource<>(entityTags, TagType.ENTITIES);
+				tagSourceMap.put(TagType.ENTITIES, paperEntityTags);
+			} catch (IllegalAccessException ignored) {}
+		}
+
+		tagSourceMap.put(TagType.ITEMS, SkriptTagSource.ITEMS);
+		tagSourceMap.put(TagType.BLOCKS, SkriptTagSource.BLOCKS);
+		tagSourceMap.put(TagType.ENTITIES, SkriptTagSource.ENTITIES);
+	}
+
+	/**
+	 * Gets all the tags of a specific origin that are applicable to a given class.
+	 * @param origin The origin to filter by.
+	 * @param typeClass The class the tags should be applicable to.
+	 * @param types Tag types to check with. Leaving this empty will check all tag types.
+	 * @return TagRegistry from the given origin and types that apply to the given class.
+	 * @param <T> see typeClass.
+	 */
+	public <T extends Keyed> Iterable<Tag<T>> getTags(TagOrigin origin, Class<T> typeClass, TagType<?>... types) {
+		List<Iterator<Tag<T>>> tagIterators = new ArrayList<>();
+		if (types == null)
+			types = tagSourceMap.map.keys().toArray(new TagType[0]);
+		for (TagType<?> type : types) {
+			if (typeClass.isAssignableFrom(type.type())) {
+				//noinspection unchecked
+				Iterator<Tag<T>> iterator = getTags(origin, (TagType<T>) type).iterator();
+				if (iterator.hasNext())
+					tagIterators.add(iterator);
+			}
+		}
+		return new Iterable<>() {
+			@Override
+			public @NotNull Iterator<Tag<T>> iterator() {
+				return Iterators.concat(tagIterators.iterator());
+			}
+		};
+	}
+
+	/**
+	 * Gets all the tags of a specific origin that are of a specific type.
+	 * @param origin The origin to filter by.
+	 * @param type The type of tags to get.
+	 * @return TagRegistry from the given origin that are of the given type.
+	 * @param <T> The class these tags apply to.
+	 */
+	public <T extends Keyed> Iterable<Tag<T>> getTags(TagOrigin origin, TagType<T> type) {
+		if (!tagSourceMap.containsKey(type))
+			return List.of();
+		Iterator<TagSource<T>> tagSources = tagSourceMap.get(origin, type).iterator();
+		if (!tagSources.hasNext())
+			return List.of();
+		return new Iterable<>() {
+			@Override
+			public @NotNull Iterator<Tag<T>> iterator() {
```

Comment:
Would it be possible to use the same `Iterators.concat` method from above here?

Reply:
good call